File: Algorithms/62. Unique Paths.rs
impl Solution {
    pub fn unique_paths(m: i32, n: i32) -> i32 {
//         if m == 1 || n == 1 {
//             return 1;
//         }
        
//         Self::unique_paths(m - 1, n) + Self::unique_paths(m, n -1)
        
        let m = m as usize;
        let n = n as usize;
        let mut res = vec![vec![0; n]; m];
        for i in 0..m {
            for j in 0..n {
                match (i, j) {
                    (0, 0) => res[i][j] = 1,
                    (0, j) => res[i][j] = res[i][j - 1],
                    (i, 0) => res[i][j] = res[i - 1][j],
                    (i, j) => res[i][j] = res[i - 1][j] + res[i][j - 1],
                }
            }
        }
        res[m - 1][n - 1]
    }
}

File: Algorithms/62. Unique Paths.dart
class Solution {
  int uniquePaths(int m, int n) {
    List<List<int>> dp = new List.generate(m, (_) => new List.generate(n, (_) => 1));
    for (var i = 1; i < m; i++) {
      for (var j = 1; j < n; j++) {
        dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
      }
    }
    return dp[m - 1][n - 1];
  }
}